### PR TITLE
Make Casper Functions Pure

### LINKED ISSUE #663: Make Functions in Casper Package Pure
A few of the functions in the casper package modify balances or change state directly instead of returning the object when the function is called. This will lead to inconsistent state being persisted if the node shuts down abruptly during a state transition. A PR for this would make the functions pure and lead it to return the object.

For Example:

The `ApplyCrosslinkRewardsAndPenalties` function in the casper package is not pure and modifies the validator balances directly instead of returning them. This would lead to balances being in an inconsistent state if `updateHead` fails after `newStateRecalc` has been run. 

Solutions to this would be to deep copy the whole of the crystallized state at the start of `newStateRecalc` or only modify the crystallized state at the end of the method.

### PR DESCRIPTION
**Pull request by IvanTheGreatDev**
Resolves #663 

This is for #663 to prevent any side effects that may occur from keeping them non-pure. If there are issues with my implementation please let me know :)

### CODE REVIEW

File: beacon-chain/casper/incentives.go [hunk 1]
```diff
@@ -18,7 +18,8 @@ func CalculateRewards(
 	voterIndices []uint32,
 	validators []*pb.ValidatorRecord,
 	totalParticipatedDeposit uint64,
-	timeSinceFinality uint64) []*pb.ValidatorRecord {
+	timeSinceFinality uint64) (newValidators []*pb.ValidatorRecord) {
```
Comment (nisdas) [inline beacon-chain/casper/incentives.go]: so this does not make it pure , you are still modifying the same underlying validator record. Assigning to a new variable does not change that

File: beacon-chain/casper/incentives.go [hunk 2]
```diff
@@ -100,16 +100,18 @@ func QuadraticPenalty(numberOfSlots uint64) uint64 {
 
 // RewardValidatorCrosslink applies rewards to validators part of a shard committee for voting on a shard.
 // TODO(#538): Change this to big.Int as tests using 64 bit integers fail due to integer overflow.
-func RewardValidatorCrosslink(totalDeposit uint64, participatedDeposits uint64, rewardQuotient uint64, validator *pb.ValidatorRecord) {
-	currentBalance := int64(validator.Balance)
-	currentBalance += int64(currentBalance) / int64(rewardQuotient) * (2*int64(participatedDeposits) - int64(totalDeposit)) / int64(totalDeposit)
-	validator.Balance = uint64(currentBalance)
+func RewardValidatorCrosslink(totalDeposit uint64, participatedDeposits uint64, rewardQuotient uint64, validator *pb.ValidatorRecord) (newValidator *pb.ValidatorRecord) {
+	newValidator = validator
```
Comment (nisdas) [inline beacon-chain/casper/incentives.go]: Same over here, you are referencing the same validator record with `newValidator`

File: beacon-chain/casper/incentives.go [hunk 3]
```diff
@@ -100,16 +100,18 @@ func QuadraticPenalty(numberOfSlots uint64) uint64 {
...
+	currentBalance := newValidator.Balance
+	newValidator.Balance += uint64(int64(currentBalance) / int64(rewardQuotient) * (2*int64(participatedDeposits) - int64(totalDeposit)) / int64(totalDeposit))
+	return newValidator
 }
 
 // PenaliseValidatorCrosslink applies penalties to validators part of a shard committee for not voting on a shard.
-func PenaliseValidatorCrosslink(timeSinceLastConfirmation uint64, rewardQuotient uint64, validator *pb.ValidatorRecord) {
-	newBalance := validator.Balance
+func PenaliseValidatorCrosslink(timeSinceLastConfirmation uint64, rewardQuotient uint64, validator *pb.ValidatorRecord) (newValidator *pb.ValidatorRecord) {
+	newValidator = validator
```
Comment (nisdas) [inline beacon-chain/casper/incentives.go]: same here

File: beacon-chain/casper/state_transition.go [hunk 4]
```diff
@@ -70,8 +70,8 @@ func ApplyCrosslinkRewardsAndPenalties(
 	attestation *pb.AggregatedAttestation,
 	validators []*pb.ValidatorRecord,
 	totalBalance uint64,
-	voteBalance uint64) error {
-
+	voteBalance uint64) ([]*pb.ValidatorRecord, error) {
+	newValidators := validators
```
Comment (nisdas) [inline beacon-chain/casper/state_transition.go]: same here

File: beacon-chain/casper/validator.go [hunk 5]
```diff
@@ -257,8 +257,8 @@ func AddPendingValidator(
 	pubKey []byte,
 	withdrawalShard uint64,
 	withdrawalAddr []byte,
-	randaoCommitment []byte) []*pb.ValidatorRecord {
-
+	randaoCommitment []byte) (newValidators []*pb.ValidatorRecord) {
+	newValidators = validators
```
Comment (nisdas) [inline beacon-chain/casper/validator.go]: same here

File: beacon-chain/casper/incentives.go (hunk 1 above)
Comment (terenc3t) [reply beacon-chain/casper/incentives.go]: I think this needs to be a deep copy?

**Pull request closed without merging**